docs: Jenny example

xjyribro opened the pull request:
<!-- Exclude from commit message -->
<!--
The title of your PR on the line above should start with a [Conventional Commit] prefix
(`fix:`, `feat:`, `docs:`, `test:`, `chore:`, `refactor:`, `perf:`, `build:`, `ci:`,
`style:`, `revert:`). This title will later become an entry in the [CHANGELOG], so please
make sure that it summarizes the PR adequately.

Don't remove the "exclude from commit message" comments below. They are used to prevent
the PR description template from being included in the git log.
Only change the "Replace this text" parts.
-->

# Description
<!--
Provide a description of what this PR is doing.
If you're modifying existing behavior, describe the existing behavior, how this PR is changing it,
and what motivated the change. If this is a breaking change, specify explicitly which APIs were
changed.
-->
<!-- End of exclude from commit message -->
Replace this text.

<!-- Exclude from commit message -->
## Checklist
<!--
Before you create this PR confirm that it meets all requirements listed below by checking the
relevant checkboxes with `[x]`. If some checkbox is not applicable, mark it as `[-]`.
-->

- [x] I have followed the [Contributor Guide] when preparing my PR.
- [x] I have updated/added tests for ALL new/updated/fixed functionality.
- [ ] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [ ] I have updated/added relevant examples in `examples` or `docs`.


## Breaking Change?
<!--
Would your PR require Flame users to update their apps following your change?

If yes, then the title of the PR should include "!" (for example, `feat!:`, `fix!:`). See
[Conventional Commit] for details. Also, for a breaking PR uncomment and fill in the "Migration
instructions" section below.
-->

- [ ] Yes, this PR is a breaking change.
- [x] No, this PR is not a breaking change.

<!--
### Migration instructions

If the PR is breaking, uncomment this header and add instructions for how to migrate from the
currently released version in-between the two following tags:
-->
<!-- End of exclude from commit message -->
<!-- Exclude from commit message -->

## Related Issues
<!--
Indicate which issues this PR resolves, if any. For example:

Closes #1234
!-->

<!-- Links -->
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Conventional Commit]: https://conventionalcommits.org
[CHANGELOG]: https://github.com/flame-engine/flame/blob/main/CHANGELOG.md
<!-- End of exclude from commit message -->

xjyribro commented:
@spydon 
This PR is for this [issue](https://github.com/flame-engine/flame/issues/2984). 

I wanted to get your feedback on this since this is the first time I'm contributing. I have done a simple Jenny example (advanced_jenny.dart is not ready). Could you take a look please. Thanks!

spydon commented on packages/flame_jenny/example/.gitignore:
```diff
@@ -0,0 +1,43 @@
+# Miscellaneous
```
This file should not be included, there is already a gitignore file in the root of the monorepo handling it.

xjyribro replied:
Removed in [bbf0135](https://github.com/flame-engine/flame/pull/3086/commits/bbf01353a5a89312cd105136eabbf20a2e366a7d)

spydon commented on packages/flame_jenny/example/README.md:
```diff
@@ -0,0 +1,16 @@
+# jenny_example
+
+Example of Jenny API
```
```suggestion
Example of how to use the Jenny package.
```

spydon commented on packages/flame_jenny/example/README.md:
```diff
@@ -0,0 +1,16 @@
+
+Example of Jenny API
+
+## Getting Started
+
+This project is a starting point for a Flutter application.
+
+A few resources to get you started if this is your first Flutter project:
+
+- [Lab: Write your first Flutter app](https://docs.flutter.dev/get-started/codelab)
+- [Cookbook: Useful Flutter samples](https://docs.flutter.dev/cookbook)
+
+For help getting started with Flutter development, view the
+[online documentation](https://docs.flutter.dev/), which offers tutorials,
+samples, guidance on mobile development, and a full API reference.
```
```suggestion
```

spydon commented on packages/flame_jenny/example/analysis_options.yaml:
```diff
@@ -0,0 +1,28 @@
+# This file configures the analyzer, which statically analyzes Dart code to
```
This should look like the analysis file of all other examples, i.e. use flame_lint

xjyribro replied:
removed in bbf01353a5a89312cd105136eabbf20a2e366a7d

xjyribro replied:
Changed lint and ran `melos run analyze`. Changes made in [480ce8e](https://github.com/flame-engine/flame/pull/3086/commits/480ce8e95b56137db2088fbcc27da9966b6f54d1)

spydon commented on packages/flame_jenny/example/assets/images/green_button.png:
These images look a bit pixelated, is it possible to upload them in higher quality?

xjyribro replied:
Actually this image was for a pixel art game so it should look a bit pixelated. I changed the image to a square instead: 
ea4af57

spydon commented on packages/flame_jenny/example/lib/components/button_row.dart:
```diff
@@ -0,0 +1,71 @@
+
+  void showNextButton(Function onNextButtonPressed) {
+    removeButtons();
+    DialogueButton nextButton = DialogueButton(
```
```suggestion
    final nextButton = DialogueButton(
```
Once you add `flame_lint` it will come with some suggestions like this.

spydon commented on packages/flame_jenny/example/lib/components/dialogue_button.dart:
```diff
@@ -0,0 +1,46 @@
+
+class DialogueButton extends SpriteButtonComponent {
+  DialogueButton({
+    required Vector2 posit,
```
```suggestion
    required super.position,
```

spydon commented on packages/flame_jenny/example/lib/components/dialogue_button.dart:
```diff
@@ -0,0 +1,46 @@
+class DialogueButton extends SpriteButtonComponent {
+  DialogueButton({
+    required Vector2 posit,
+    required String assetPath,
+    required String text,
+    required super.onPressed,
+    super.anchor = Anchor.center,
+  }) : super() {
+    position = posit;
+    _text = text;
+    _assetPath = assetPath;
+  }
+
+  late String _text;
+  late String _assetPath;
```
```suggestion
    required this.assetPath,
    required this.text,
    required super.onPressed,
    super.anchor = Anchor.center,
  });

  final String text;
  final String assetPath;
```

spydon commented on packages/flame_jenny/example/lib/components/dialogue_button.dart:
```diff
@@ -0,0 +1,46 @@
+
+  @override
+  Future<void> onLoad() async {
+    Sprite dButton = await Sprite.load(_assetPath);
+    button = dButton;
```
```suggestion
    button = await Sprite.load(_assetPath);
```
No abbreviations are allowed in this repository

spydon commented on packages/flame_jenny/example/lib/components/dialogue_text_box.dart:
```diff
@@ -0,0 +1,18 @@
+import 'package:jenny_example/commons/commons.dart';
+
+class DialogueTextBox extends TextBoxComponent {
+  DialogueTextBox({required String text})
+      : super(
+          text: text,
```
```suggestion
  DialogueTextBox({required super.text})
      : super(
```

spydon commented on packages/flame_jenny/example/lib/main.dart:
```diff
@@ -0,0 +1,19 @@
+  runAsDashbook();
+}
+
+void runAsDashbook() {
+  final dashbook = Dashbook(
+    title: 'Jenny Examples',
+  );
+
+  addJennySimpleExample(dashbook);
+  addJennyAdvancedExample(dashbook);
+
+  runApp(dashbook);
+}
```
Just keep the simple example in here and no dashbook, and have both examples inside of the root examples directory instead, create a directory called `flame_jenny` here and follow the same structure as the other examples:
https://github.com/flame-engine/flame/tree/main/examples/lib/stories/bridge_libraries

xjyribro replied:
Moved in [721959c](https://github.com/flame-engine/flame/pull/3086/commits/721959c28e1dc4b92c5ecf5d4dfde0556e672d05)

spydon commented on packages/flame_jenny/example/lib/stories/advanced_jenny.dart:
```diff
@@ -0,0 +1,25 @@
+
+class JennyAdvancedExample extends FlameGame {
+  static const String description = '''
+    This example shows how to use the Jenny API. .
```
```suggestion
    This example shows how to use the Jenny package. .
```

spydon commented on packages/flame_jenny/example/pubspec.yaml:
```diff
@@ -0,0 +1,31 @@
+name: jenny_example
+description: Example of Jenny API
```
```suggestion
description: An example of the Jenny package
```

spydon commented on packages/flame_jenny/example/pubspec.yaml:
```diff
@@ -0,0 +1,31 @@
+  flutter_test:
+    sdk: flutter
+
+  flutter_lints: ^3.0.0
```
```suggestion
  flame_lint: ^1.1.2
```

spydon reviewed (commented):
The logic looks good to me, some comments about the structure though.

xjyribro commented:
@spydon 
I have made the changes according to your feedback. Summary of changes:

1. Moved the folder to https://github.com/flame-engine/flame/tree/main/examples/lib/stories/bridge_libraries/flame_jenny
2. Changed `analysis_options.yaml` and fixed code after lint check.
3. Removed .gitignore

Other updates:

1. Added advanced example. **Please give me your feedback for this portion.** I did not include all of Jenny Package's features, only selected ones. If there are features that should be included, I can add them.
2. Updated the descriptions.
3. I have not changed the button sprite image yet. Working on it.

spydon commented on examples/lib/stories/bridge_libraries/flame_jenny/jenny.dart:
```diff
@@ -0,0 +1,29 @@
+import 'package:flame/game.dart';
+
+void addJennyExample(Dashbook dashbook) {
+  dashbook.storiesOf('FlameJenny')
```
Make sure this follows the same standard as the other bridge libraries

xjyribro replied:
I think this is what you mean. [b76db82](https://github.com/flame-engine/flame/pull/3086/commits/b76db828df4e21a13cb988f560e1f338c23091a4)

spydon replied:
I meant this place, but I see now that this is inline with the others and that it's flame_forge2d that is deviating from the standard

spydon reviewed (commented):
Good job! I don't think you have to add anymore features to the advanced example, this is a good starting point for users.
Just a few minor comments, then I think it's ready to get merged. :)

xjyribro commented on examples/lib/stories/bridge_libraries/flame_jenny/components/dialogue_button.dart:
```diff
@@ -0,0 +1,41 @@
+    super.anchor = Anchor.center,
+  });
+
+  final String text;
```
This button is used for both going to the next line and options in dialgoues. For simple dialogues with no options, I can simply change the text like you mentioned, but if I have options I still need to remove and add the buttons for options then remove the option buttons and add the next button again. I feel this way is cleaner, less confusing.

xjyribro commented on examples/lib/stories/bridge_libraries/flame_jenny/components/dialogue_box.dart:
```diff
@@ -0,0 +1,54 @@
+
+  void changeText(String newText, Function() goNextLine) {
+    removeTextBox();
+    textBox = DialogueTextBox(text: newText);
```
changed in [f2ee10a](https://github.com/flame-engine/flame/pull/3086/commits/f2ee10af4c5e74febe40d09bc9dcfe9c786f9f7e)

xjyribro commented on examples/lib/stories/bridge_libraries/flame_jenny/components/menu_button.dart:
```diff
@@ -0,0 +1,39 @@
+  final TextPaint topTextPaint = TextPaint(
+    style: TextStyle(color: BasicPalette.black.color),
+  );
+  final startButtonSize = Vector2(128, 42);
```
set in [a71ec34](https://github.com/flame-engine/flame/pull/3086/commits/a71ec34b9e2331fdcbade7ce430d50159ec69179)

spydon reviewed (approved):
Lgtm, thanks for your contribution!
